src/relay/backend/compile_engine.cc

```diff
@@ -593,8 +604,10 @@ class MakeShapeFunc : public backend::MemoizedExprTranslator<Array<te::Tensor>>
   std::unordered_map<Expr, Array<te::Tensor>, ObjectPtrHash, ObjectPtrEqual> param_data_;
   /*! \brief Map from parameter to list of shape placeholder */
   std::unordered_map<Expr, Array<te::Tensor>, ObjectPtrHash, ObjectPtrEqual> param_shapes_;
-  /*! \brief Stack of data dependencies for shape function */
+  /*! \brief Stack of data dependencies for shape function, specified per op */
   std::vector<bool> data_dependants_;
```

Comment:
I think you should remove the `data_dependants_` and only use `data_dependants_per_input_`.

Reply:
There is a usage of `data_dependants_` at https://github.com/apache/tvm/blob/f658556d5f9491138fde4451baf60c9621d0755d/src/relay/backend/compile_engine.cc#L453

I didn't look into this in detail, can we replace this with `data_dependants_per_input_`? To me it seems per-op data dependant makes more sense here.

Reply:
Correct me if I'm wrong. The `data_dependent_` logic is to determine whether we need shape or data from each node when visiting it. This applies to `Var`, `Constant`, and `Call`. Since the data dependency could be now different for each arg in the Call, we should always use `data_dependants_per_input_`.

Reply:
btw. It's a typo for `dependant`. It should be `dependent`. Could you correct it at the same time?

Reply:
ok I'll try removing `data_dependants_`.

Right, I was not sure `dependant` was intentional or not... It's actually everywhere (grep `dependant` gives many hit). I'll fix them all!!